### PR TITLE
cli: create -addrinfo

### PR DESCRIPTION
**Pull request by jonatack**
While looking at issue #21351, it turned out that the problem was a lack of tor v3 addresses known to the node. It became clear (e.g. https://github.com/bitcoin/bitcoin/issues/21351#issuecomment-811004779) that a CLI command returning the number of addresses the node knows per network (with a tor v2 / v3 breakdown) would be very helpful. This patch adds that.

`-addrinfo` is useful to see if your node knows enough addresses in a network to use options like `-onlynet=<network>`, or to upgrade to the upcoming tor release that no longer supports tor v2, for which you'll need to be sure your node knows enough tor v3 peers.

```
$ bitcoin-cli --help | grep -A1 addrinfo
  -addrinfo
       Get the number of addresses known to the node, per network and total.

$ bitcoin-cli -addrinfo
{
  "addresses_known": {
    "ipv4": 14406,
    "ipv6": 2511,
    "torv2": 5563,
    "torv3": 2842,
    "i2p": 8,
    "total": 25330
  }
}

$ bitcoin-cli -addrinfo 1
error: -addrinfo takes no arguments
```

This can be manually tested, for example, with commands like this:
```
$ bitcoin-cli getnodeaddresses 0 | jq '.[] | (select(.address | contains(".onion")) | select(.address | length <= 22)) | .address' | wc -l
5563
$ bitcoin-cli getnodeaddresses 0 | jq '.[] | (select(.address | contains(".onion")) | select(.address | length > 22)) | .address' | wc -l
2842
$ bitcoin-cli getnodeaddresses 0 | jq '.[] | .address' | wc -l
25330
```

### CODE REVIEW

**Review by 78051301012 — APPROVED** [review]

**Comment by practicalswift** [comment]
Concept ACK: nice! :)

**Comment by szeli57** [comment]
P

**Review by jarolrod — COMMENTED** [review]
ACK 5f1ffbfa3fccb145a6724ef33f8de92d6b652a31

Tested on macOS 11.2.3
```
xyz@xyzs-MBP bitcoin % ./src/bitcoin-cli -testnet -addressinfo
{
  "addresses known": {
    "ipv4": 9774,
    "ipv6": 2211,
    "torv2": 0,
    "torv3": 0,
    "i2p": 0,
    "total": 11985
  }
}
```

Question: Why make this a CLI command and not an RPC command? Seems useful as an RPC command.

**Comment by jonatack** [comment]
Thanks!

> Question: Why make this a CLI command and not an RPC command? Seems useful as an RPC command.

It could, or a third path could be to add a boolean `totals` arg to the `getnodeaddresses` RPC to toggle a "totals mode" (that said, it might be a bit obscure/hard to discover).

Generally, CLI commands batch or wrap one or multiple RPC commands and are easier to change and improve without worrying about API stability. For example, this might evolve to returning other address info or statistics. In this case where only one RPC is wrapped, instead of batching multiple RPCs like `-getinfo`, `-netinfo` and `-generate` do, I don't have a strong opinion on the RPC vs CLI question.

**Comment by jonatack** [comment]
Not sure, but it's maybe more flexible to start life as a CLI and then become an RPC if there's a good reason, than vice-versa.

**Comment by laanwj** [comment]
Concept ACK

> Question: Why make this a CLI command and not an RPC command? Seems useful as an RPC command.

Minimalism. I think it's a good choice to put functionality in the client that can be done client-side. Adding things server-side adds maintenance burden, means having to commit to a certain (machine parseable) interface, etc. And this is clearly functionality aimed at end users. Also agree with

> Not sure, but it's maybe more flexible to start life as a CLI and then become an RPC if there's a good reason, than vice-versa.

**Comment by jonatack** [comment]
An argument I could see for an RPC would be potentially better performance by avoiding a copy of a large object and by parsing Network enums rather than strings. Doesn't seem critical for occasional use though.

**Comment by jarolrod** [comment]
@laanwj @jonatack Thanks for answering my questions. Maybe this information, when a command should be a CLI or RPC command, should be added to the [developer-notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#rpc-interface-guidelines).

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -569,6 +569,56 @@ class NetinfoRequestHandler : public BaseRequestHandler
...
+            ++m_counts.at(network_id);
+        }
+        // Prepare result to return to user.
+        UniValue result{UniValue::VOBJ}, addresses{UniValue::VOBJ};
+        for (uint8_t i = 0; i < m_networks_size; ++i) {
+            addresses.pushKV(m_networks.at(i), m_counts.at(i));
+            m_total += m_counts.at(i);
+        }
+        addresses.pushKV("total", m_total);
+        result.pushKV("addresses known", addresses);
```
Comment (promag) [inline src/bitcoin-cli.cpp:618]: 57eb71ce0f27c0353ea0742a696c6d04ce048e15

IMO avoid spaces in keys.

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -569,6 +569,56 @@ class NetinfoRequestHandler : public BaseRequestHandler
...
 };
 
+/** Process addressinfo requests */
+class AddressinfoRequestHandler : public BaseRequestHandler
+{
+private:
+    static constexpr uint8_t m_networks_size{5};
+    const std::array<std::string, m_networks_size> m_networks{{"ipv4", "ipv6", "torv2", "torv3", "i2p"}};
+    std::array<uint64_t, m_networks_size> m_counts{{}}; // Address counts by network
+    uint64_t m_total{0}; // Total address count
```
Comment (promag) [inline src/bitcoin-cli.cpp:580]: 57eb71ce0f27c0353ea0742a696c6d04ce048e15

No need to have `m_networks` and `m_total` as members.

File: src/bitcoin-cli.cpp [hunk 3]
```diff
@@ -569,6 +569,56 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "> bitcoin-cli -netinfo help\n"};
 };
 
+/** Process addressinfo requests */
+class AddressinfoRequestHandler : public BaseRequestHandler
+{
+private:
+    static constexpr uint8_t m_networks_size{5};
```
Comment (promag) [inline src/bitcoin-cli.cpp:577]: 57eb71ce0f27c0353ea0742a696c6d04ce048e15

Drop, and use `m_networks` size?

File: src/bitcoin-cli.cpp [hunk 4]
```diff
@@ -569,6 +569,56 @@ class NetinfoRequestHandler : public BaseRequestHandler
...
+    const std::array<std::string, m_networks_size> m_networks{{"ipv4", "ipv6", "torv2", "torv3", "i2p"}};
```
Comment (promag) [inline src/bitcoin-cli.cpp:578]: 57eb71ce0f27c0353ea0742a696c6d04ce048e15

I think you can omit template arguments, see https://stackoverflow.com/a/61371266/1978589.

**Review by promag — CHANGES_REQUESTED** [review]
Code review ACK 5f1ffbfa3fccb145a6724ef33f8de92d6b652a31 with some suggestions.

But I'm not fond of this approach. This is fetching a list from the server and doing an aggregation on the client, this can be done by other means. I would be in favor of implementing this on the server so we get it right for all clients, GUI included.

I'm also not fond of the name `-addressinfo` as it looks like something specific to an address only, I would name it `-nodestats` or something like that.

File: src/bitcoin-cli.cpp (hunk 4 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp:578]: Thanks, will give it a try in the next push.

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp:618]: Good point, will snakecase it if this remains in JSON format. I wonder if it should drop the JSON format (like `-netinfo`) for the client side.

**Comment by jonatack** [comment]
> I would be in favor of implementing this on the server so we get it right for all clients, GUI included.
 
The GUI is a good point, and performance could also be better (not sure if it would be noticeable) as an RPC.

> I'm also not fond of the name `-addressinfo` as it looks like something specific to an address only, I would name it `-nodestats` or something like that.

I proposed `-addressinfo` in line with CLI commands `-getinfo` and `-netinfo`...`nodestats` isn't bad though and is shorter. Hm...

**Comment by jonatack** [comment]
Note to self: as a CLI this needs a server version check like the one in `-netinfo` to gracefully raise if the server is earlier than v22.

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp:618]: done

File: src/bitcoin-cli.cpp (hunk 2 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp:580]: Changed `m_counts` and `m_total` to be local variables instead (`counts` and `total`).

File: src/bitcoin-cli.cpp (hunk 4 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp:578]: Couldn't get that to compile (clang 11).

**Comment by jonatack** [comment]
- Rebased following the merge of #21594
- Took review feedback where applicable
- Re "-nodestats", I couldn't get any command name beginning with "-node" to work (e.g. `nodestats`, `nodeinfo`), it compiles, but when run: "Error parsing command line arguments: Invalid parameter -nodestats"...not sure why, verboten command name?
- Shortened the command from `-addressinfo` to `-addrinfo`, as shorter is better for humans typing manually
- Added the server version check which raises with a helpful message
- Improved the release note

File: src/bitcoin-cli.cpp (hunk 3 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp]: I would if could omit the template arguments.

**Review by jarolrod — COMMENTED** [review]
tACK c0986af3c1a7eba860f9375674f03aaa365e4f72

Tested on Ubuntu and macOS 11.2.3

I think the change from `-addressinfo` to `-addrinfo` is more appropriate as `-addressinfo` does sound like you are trying to get information about a specific address.

one NIT: 
The behavior when the server is initiating is a bit different than other CLI commands like `-getinfo` and `-netinfo`. When you call either of these while the server is starting up you'll get something like this:

```
error code: -28
error message:
Loading block index...
```

When calling this command, `-addrinfo`,  while the server is starting you'll get an error about the output:

```
error: JSON value is not an object or array as expected
```

I don't know how important this is, but maybe we should keep this behavior consistent?

**Comment by jonatack** [comment]
Good find! I agree this isn't great, will look at fixing.

**Comment by jonatack** [comment]
Thanks again for finding that @jarolrod. Fixed in the last commit per `git diff c0986af edf3167`. Much better.

```diff
@@ -256,6 +256,7 @@ public:
 
     UniValue ProcessReply(const UniValue& reply) override
     {
+        if (!reply["error"].isNull()) return reply;
         const std::vector<UniValue>& nodes{reply["result"].getValues()};
```
```
$ bitcoin-cli -addrinfo
error code: -28
error message:
Loading block index...
````

**Comment by laanwj** [comment]
> I would be in favor of implementing this on the server so we get it right for all clients, GUI included.

I'm not 100% convinced this is worth adding server side. Also not against it, but I don't think we should add a lot of "would be nice" RPC commands. Yes, doing it client side is slower but how often are you going to call this ?

**Comment by jonatack** [comment]
Right, and the fields returned here will probably change...more BIP155 networks may be on the way...and we may want to replace "torv2" and "torv3" with "onion" a year from now...or maybe add "torv4"...easier to do those updates as a CLI.

**Review by jarolrod — COMMENTED** [review]
tACK edf3167151f7a6d08cf733b4e230e2d745819ac8

Server starting up
```
error code: -28
error message:
Loading block index...
```
Server Running
```
{
  "addresses_known": {
    "ipv4": 15227,
    "ipv6": 3583,
    "torv2": 0,
    "torv3": 0,
    "i2p": 0,
    "total": 18810
  }
}
```

Server shutting down, this message is consistent across other CLI commands

```
error: Could not connect to the server 127.0.0.1:18332

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.
```

Tested network message with this [patch](https://github.com/bitcoin/bitcoin/pull/21595#issuecomment-818156765)
```
error: -addrinfo requires bitcoind server to be running v22.0 and up
```

**Comment by jonatack** [comment]
I'm at **2842** torv3 peers known now, up from ~1800 ten days ago and still finding more...good to see.
```
{
  "addresses_known": {
    "ipv4": 14398,
    "ipv6": 2509,
    "torv2": 5561,
    "torv3": 2842,
    "i2p": 8,
    "total": 25318
  }
}
```

**Comment by jonatack** [comment]
And one more i2p peer as well!

File: src/bitcoin-cli.cpp [hunk 5]
```diff
@@ -228,6 +230,61 @@ class BaseRequestHandler
     virtual UniValue ProcessReply(const UniValue &batch_in) = 0;
 };
 
+/** Process addrinfo requests */
+class AddrinfoRequestHandler : public BaseRequestHandler
+{
+private:
+    static constexpr uint8_t m_networks_size{5};
+    const std::array<std::string, m_networks_size> m_networks{{"ipv4", "ipv6", "torv2", "torv3", "i2p"}};
```
Comment (promag) [inline src/bitcoin-cli.cpp:238]: Does the following work?
```cpp
static constexpr std::array m_networks{"ipv4", "ipv6", "torv2", "torv3", "i2p"};
```

Comment (jonatack) [reply src/bitcoin-cli.cpp]: Thanks! I think I hadn't dropped the type specifier, only the size arg. Took your suggestions and updated the file in 06c43201a714b0426cc68b2fd5c681e5df10af99.

Comment (promag) [reply src/bitcoin-cli.cpp]: Cool 👍

**Comment by laanwj** [comment]
Tested ACK 06c43201a714b0426cc68b2fd5c681e5df10af99
```javascript
% cli -addrinfo
{
  "addresses_known": {
    "ipv4": 52404,
    "ipv6": 10827,
    "torv2": 4987,
    "torv3": 94,
    "i2p": 5,
    "total": 68317
  }
}
```
